### PR TITLE
Setup Email Notifications

### PR DESCRIPTION
**Pull request by areetaw**
### Email Prototype

**OAuth 2.0 Client ID**
To access protected data and get authorized into someone's Gmail account, I added the OAuth consent screen with specifically the `https://mail.google.com/` access scope which allows reading, composing, sending, and deleting emails. Currently, this gives noms all access to the users email for testing purposes but will switch to only sending in the future. Then, I created the OAuth 2.0 Client IDs including all credentials needed to access the Gmail API. I added the `credentials.json` file to `.gitignore` to prevent outsiders from access. 

**OAuth 2.0 Accessing Gmail API**
In order to access APIs, the Google Authorization Server must give access and refresh tokens; this is retrieved through a GET and POST method from the user. Since noms will be sending notifs internally, I used my own corporate Google email and initiated the authentication and authorization process through redirecting to:

`https://accounts.google.com/o/oauth2/v2/auth?
 scope=https://mail.google.com/&
 access_type=offline&
 include_granted_scopes=true&
 response_type=code&
 state=state_parameter_passthrough_value&
 redirect_uri=http://localhost:8080&
 client_id=client_id` 

and accepting the Google prompted user consent form. This gives me an authorization code to make a POST request: 

 `curl \
--request POST \
--data "code=[Authentcation code from authorization link]&client_id=[Application Client Id]&client_secret=[Application Client Secret]&redirect_uri=http://localhost&grant_type=authorization_code" \
https://accounts.google.com/o/oauth2/token`

and returns: 

`{
  "access_token": "1/fFAGRNJru1FTz70BzhT3Zg",
  "expires_in": 3920,
  "token_type": "Bearer",
  "scope": "https://mail.google.com/",
  "refresh_token": "1//xEoDL4iW3cxlI7yDbSRFYNG01kVKM2C-259HOF2aQbI"
}`

**Gmail API Usage and Gmail Configuration**
After gathering client information and tokens, I created the Gmail service to access the API. Then, I included a `getAccessToken(String refresh_token)` to update the access token since it only has a lifetime of 1 hour whenever service is needed. Additionally, I was able to send a test email to myself through `MimeMessage`, a Java email-like message and Gmail's `Message`.

**Example of Test Email**
image

### CODE REVIEW

File: src/main/java/com/google/sps/api/GmailAPI.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+	private static final String USER = "me";
+  private static final File CREDENTIALS_FILE_PATH = new File(System.getProperty("user.dir") + "/credentials.json");
+  
+  private static String CLIENT_ID = "";
+  private static String CLIENT_SECRET = "";
+	private static Gmail service = null;
+
+  /**
+    * Create a Gmail service using credentails.
```
Comment (mirrorkeydev) [inline src/main/java/com/google/sps/api/GmailAPI.java:57]: *credentails -> credentials

File: pom.xml [hunk 2]
```diff
@@ -104,6 +104,31 @@
...
+    <dependency>
+      <groupId>commons-codec</groupId>
+      <artifactId>commons-codec</artifactId>
+      <version>1.14</version>
+    </dependency>
+    <dependency>
+			<groupId>com.google.apis</groupId>
+			<artifactId>google-api-services-gmail</artifactId>
+			<version>v1-rev33-1.20.0</version>
+		</dependency>
```
[the remark this answers is not in the document]
Comment (mirrorkeydev) [reply pom.xml:121]: The other possibility is that if you copy and pasted code from the docs, for example, it may have used tabs instead of spaces and they didn't get automatically converted

File: src/main/java/com/google/sps/api/GmailConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    * Send a test email from areeta@google.com.
+    *
+    * @throws IOException
+    * @throws GeneralSecurityException
+    * @throws MessagingException
+    */
+  public static void sendEmail() 
+      throws IOException, GeneralSecurityException, MessagingException {
+		Gmail service = GmailAPI.getGmailService();
+		MimeMessage Mimemessage = createEmail("areeta@google.com","me","This my demo test subject","This is my body text");
```
Comment (ubahl) [inline src/main/java/com/google/sps/api/GmailConfiguration.java:119]: In the future you could make variables (ex String email = areeta@google.com, etc..) and then put those variables in the createEmail function instead of the strings themselves.

File: pom.xml [hunk 4]
```diff
@@ -104,6 +104,31 @@
       <artifactId>slf4j-simple</artifactId>
       <version>1.7.25</version>
     </dependency>
+    <dependency>
+			<groupId>org.json</groupId>
+			<artifactId>json</artifactId>
+			<version>20190722</version>
+		</dependency>
+    <dependency>
```
[the remark this answers is not in the document]
Comment (areetaw) [reply pom.xml:112]: Just checked and it's GitHub being weird!

File: pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (areetaw) [reply pom.xml]: My cloud shell is configured to 2 spaces as tabs! The other possibility with copy/pasted dependencies is the issue; I just went ahead and updated the formatting. Good catch!

File: src/main/java/com/google/sps/api/GmailConfiguration.java (hunk 3 above)
Comment (areetaw) [reply src/main/java/com/google/sps/api/GmailConfiguration.java:119]: Over here, I included as everything as strings just to test it and plan to use variables in the future!

File: src/main/java/com/google/sps/api/GmailAPI.java (hunk 1 above)
Comment (areetaw) [reply src/main/java/com/google/sps/api/GmailAPI.java:57]: Great catch!

File: src/main/java/com/google/sps/api/GmailAPI.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+				buffer.append(line);
+			}
+
+      // Return select access token from POST response.
+			JSONObject json = new JSONObject(buffer.toString());
+			String accessToken = json.getString("access_token");
+			return accessToken;
+		
+    } catch (Exception ex) {
+			ex.printStackTrace();
```
Comment (mrboxtobox) [inline src/main/java/com/google/sps/api/GmailAPI.java:137]: Explain why you only print the stacktrace e.g. why it's okay to ignore the error.

File: src/main/java/com/google/sps/api/GmailAPI.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    CLIENT_ID = clientSecrets.getDetails().getClientId().toString();
+    CLIENT_SECRET = clientSecrets.getDetails().getClientSecret().toString();
+
+    // Set up credentials to use the Gmail API.
+		Credential authorize = new GoogleCredential.Builder().setTransport(GoogleNetHttpTransport.newTrustedTransport())
+				.setJsonFactory(JSON_FACTORY)
+				.setClientSecrets(CLIENT_ID, CLIENT_SECRET)
+				.build()
+        .setRefreshToken(Token.REFRESH_TOKEN)
+        .setAccessToken(getAccessToken(Token.REFRESH_TOKEN));
```
Comment (mrboxtobox) [inline src/main/java/com/google/sps/api/GmailAPI.java:76]: Since it is possible to return `null` from `getAccessToken`, I would like to see some error handling done. A TODO is fine for now.

File: src/main/webapp/about.html [hunk 7]
```diff
@@ -5,11 +5,11 @@
       <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no, minimal-ui">
       <title>about</title>
       <!-- css -->
-      <link type="text/css" rel="stylesheet" href="./assets/css/mainStyle.css">
```
Comment (mrboxtobox) [inline src/main/webapp/about.html:8]: Why is this removed?

Comment (areetaw) [reply src/main/webapp/about.html:8]: For this PR, I accidentally included commits that weren't merged to master yet so I reverted them here.

File: src/main/java/com/google/sps/api/GmailAPI.java (hunk 6 above)
Comment (areetaw) [reply src/main/java/com/google/sps/api/GmailAPI.java:76]: Gotcha, will do!

File: src/main/webapp/about.html (hunk 7 above)
Comment (mirrorkeydev) [reply src/main/webapp/about.html:8]: Commits from a different PR that hadn't been merged yet? :0

File: src/main/java/com/google/sps/api/GmailAPI.java (hunk 5 above)
Comment (areetaw) [reply src/main/java/com/google/sps/api/GmailAPI.java:137]: If the POST request to create a new access token does not work, it will include a response with an error description so exceptions for the request wouldn't appear. Additionally, I just configured the tokens so there shouldn't be any current exceptions related to the request. In the future, I will add more exception handling!
